### ISSUE TITLE
PVFeedParser - Mutating array error happens whenever app enters the foreground

### ISSUE THREAD

**Issue by mitchdowney** [podverse/podverse#52]
According to my tests this issue goes away when I comment out lines 109 - 118 in the PVFeedParser:

```
for var existingEpisode in downloadedEpisodes {
    if newEpisode.mediaURL == existingEpisode.mediaURL {
        existingEpisode = newEpisode
        episodeAlreadySaved = true
        //Remove the created entity from core data if it already exists
        CoreDataHelper.removeManagedObjectFromClass("Episode", managedObjectContext: Constants.moc, object: newEpisode)
        break
    }
}
```

This block is needed to prevent duplicates of the same episode from being saved when the feedparser runs again.

I'm not sure why this block is causing the issue so I won't push an attempted fix for it.

But I'm thinking the .removeManagedObjectFromClass is at the root of the issue. When removeManagedObjectFromClass removes the object newEpisode, well existingEpisode and newEpisode are the exact same thing, so maybe there are two instances of the same episode floating around in the MOC, and this results in the wrong managed object getting deleted?

**Comment by mitchdowney** [comment]
Hmm, it seems like this block:

```
// If only parsing for the latest episode, stop parsing after parsing the first episode.
if shouldGetMostRecentEpisode == true {
    latestEpisodeInFeed = newEpisode
    parser.abortParsing()
}
```

Should happen before these blocks:

```
// If episode already exists in the database, do not insert new episode, instead update existing episode
for var existingEpisode in downloadedEpisodes {
    if newEpisode.mediaURL == existingEpisode.mediaURL {
        existingEpisode = newEpisode
        episodeAlreadySaved = true
        //Remove the created entity from core data if it already exists
        CoreDataHelper.removeManagedObjectFromClass("Episode", managedObjectContext: Constants.moc, object: newEpisode)
        break
    }
}

// If episode is not already saved, then add episode to the podcast object
if !episodeAlreadySaved {
    podcast.addEpisodeObject(newEpisode)
}
```

After moving the "abortParsing" block before the other two, **I no longer get the mutating array error, however I do encounter this new error:**

2015-11-27 15:27:30.141 podverse[1759:266979] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSXMLParser does not support reentrant parsing.'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000105b48f45 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x0000000107d44deb objc_exception_throw + 48
	2   CoreFoundation                      0x0000000105b48e7d +[NSException raise:format:] + 205
	3   Foundation                          0x00000001060d03d6 +[NSXMLParser setCurrentParser:] + 150
	4   Foundation                          0x00000001060d0be6 -[NSXMLParser parseFromStream] + 54
	5   podverse                            0x00000001051bc029 _TFC8podverse10FeedParser5parsefS0_FT_T_ + 2505
	6   podverse                            0x00000001051ebc82 _TFC8podverse12PVFeedParser16parsePodcastFeedfS0_FSST_ + 402
	7   podverse                            0x00000001051eea4b _TFC8podverse12PVFeedParser24feedParserParsingAbortedfS0_FCS_10FeedParserT_ + 1227
	8   podverse                            0x00000001051eecea _TToFC8podverse12PVFeedParser24feedParserParsingAbortedfS0_FCS_10FeedParserT_ + 58
	9   podverse                            0x00000001051bcc7a _TPAo16 + 42
	10  podverse                            0x00000001051d1cd9 _TTRXFo_dC8podverse10FeedParser_dT__XFo_oS0__dT__ + 25
	11  podverse                            0x00000001051d1d07 _TTRXFo_oC8podverse10FeedParser_dT__XFo_iS0__iT__ + 23
	12  podverse                            0x00000001051bcdb1 _TPA__TTRXFo_oC8podverse10FeedParser_dT__XFo_iS0__iT__ + 81
	13  podverse                            0x00000001051bcbf1 _TFC8podverse10FeedParser12abortParsingfS0_FT_T_ + 641
	14  podverse                            0x00000001051edd98 _TFC8podverse12PVFeedParser10feedParserfS0_FTCS_10FeedParser12didParseItemCS_8FeedItem_T_ + 2760
	15  podverse                            0x00000001051ee56f _TToFC8podverse12PVFeedParser10feedParserfS0_FTCS_10FeedParser12didParseItemCS_8FeedItem_T_ + 79
	16  podverse                            0x000000010516e675 _TPAo + 53
	17  podverse                            0x000000010516ed2d _TTRXFo_dC8podverse10FeedParserdCS_8FeedItem_dT__XFo_oS0_oS1__dT__ + 29
	18  podverse                            0x000000010516ed73 _TTRXFo_oC8podverse10FeedParseroCS_8FeedItem_dT__XFo_iTS0_12didParseItemS1___iT__ + 35
	19  podverse                            0x000000010516e7b1 _TPA__TTRXFo_oC8podverse10FeedParseroCS_8FeedItem_dT__XFo_iTS0_12didParseItemS1___iT__ + 81
	20  podverse                            0x000000010516cb65 _TFC8podverse16CustomFeedParser21parseEndOfRSS2ElementfS0_FTSS13qualifiedNameGSQSS__T_ + 1221
	21  podverse                            0x00000001051d3067 _TFFC8podverse10FeedParser6parserFS0_FTCSo11NSXMLParser13didEndElementSS12namespaceURIGSqSS_13qualifiedNameGSqSS__T_U_FT_T_ + 1191
	22  podverse                            0x00000001051bf7dd _TPA__TFFC8podverse10FeedParser6parserFS0_FTCSo11NSXMLParser13didEndElementSS12namespaceURIGSqSS_13qualifiedNameGSqSS__T_U_FT_T_ + 189
	23  libswiftObjectiveC.dylib            0x0000000108743bb9 _TF10ObjectiveC15autoreleasepoolFFT_T_T_ + 41
	24  podverse                            0x00000001051bf650 _TFC8podverse10FeedParser6parserfS0_FTCSo11NSXMLParser13didEndElementSS12namespaceURIGSqSS_13qualifiedNameGSqSS__T_ + 400
	25  podverse                            0x00000001051bfa12 _TToFC8podverse10FeedParser6parserfS0_FTCSo11NSXMLParser13didEndElementSS12namespaceURIGSqSS_13qualifiedNameGSqSS__T_ + 546
	26  Foundation                          0x00000001060d366b _endElementNs + 696
	27  libxml2.2.dylib                     0x000000010b36ec5a xmlParseEndTag2 + 917
	28  libxml2.2.dylib                     0x000000010b371242 xmlParseTryOrFinish + 3216
	29  libxml2.2.dylib                     0x000000010b3703bd xmlParseChunk + 905
	30  Foundation                          0x00000001060d08b5 -[NSXMLParser parseData:] + 293
	31  Foundation                          0x00000001060d0b29 -[NSXMLParser parseData:] + 921
	32  Foundation                          0x00000001060d0c83 -[NSXMLParser parseFromStream] + 211
	33  podverse                            0x00000001051bc029 _TFC8podverse10FeedParser5parsefS0_FT_T_ + 2505
	34  podverse                            0x00000001051ebc82 _TFC8podverse12PVFeedParser16parsePodcastFeedfS0_FSST_ + 402
	35  podverse                            0x0000000105193248 _TFFC8podverse11AppDelegate18refreshPodcastFeedFS0_FT_T_U_FT_T_ + 376
	36  podverse                            0x000000010517fac7 _TTRXFo__dT__XFdCb__dT__ + 39
	37  libdispatch.dylib                   0x000000010882ee5d _dispatch_call_block_and_release + 12
	38  libdispatch.dylib                   0x000000010884f49b _dispatch_client_callout + 8
	39  libdispatch.dylib                   0x0000000108837bef _dispatch_root_queue_drain + 1829
	40  libdispatch.dylib                   0x00000001088374c5 _dispatch_worker_thread3 + 111
	41  libsystem_pthread.dylib             0x0000000108b80a9d _pthread_wqthread + 729
	42  libsystem_pthread.dylib             0x0000000108b7e3dd start_wqthread + 13
)
libc++abi.dylib: terminating with uncaught exception of type NSException

**Comment by mitchdowney** [comment]
This StackOverflow post recommends creating a "reentrantAvoidanceQueue"...

http://stackoverflow.com/questions/25642070/nsxmlparser-on-ios8-nsxmlparser-does-not-support-reentrant-parsing

**Comment by mitchdowney** [comment]
The reentrantAvoidanceQueue seems to prevent the 'NSXMLParser does not support reentrant parsing.' issue.

I am pushing a fix for the Mutating Array and NSXMLParser issue to the mutatingParser branch for @kreonjr and I to review.

**Comment by kreonjr** [comment]
Closing until it shows up again. Fix in latest PR
